### ISSUE TITLE
dind in worker controller?

### ISSUE THREAD

**Issue by braun1928** [spacelift-io/spacelift-helm-charts#56]
With the new controller, `dind` container was not added, and there is just a note deprecating the old worker chart (that had `dind`). Is there a reason (aside from the usual, how bad `dind` is)?

Running the worker in k8s that might use `docker` when running a terraform file with terraform-aws-lambda module makes it needed; I kinda prefer to have Docker in Docker than exposing the host's socket, specially since it will be just a build and all get dropped afterwards.

**Comment by adamconnelly** [comment]
@braun1928 thanks a lot for bringing this up. Most of the feedback we've heard so far has been about removing the reliance on DinD, so it's really interesting to hear a use-case where you would actually like to be able to use it.

There's a big difference between how the old and new Helm charts function. With the old DinD approach, it creates a Deployment where each Pod within the Deployment run a long-running Spacelift worker process. Whereas with the new approach, the controller that's deployed as part of the Helm chart doesn't actually execute your Spacelift runs. There's a detailed explanation of how both approaches work if you're interested [on our blog](https://spacelift.io/blog/spacelift-workerpool-operator).

With the new operator approach, what happens is that a new Pod is created for each run, that lives for the lifetime of the run (with the caveat that certain runs like those that go unconfirmed will use multiple Pods). The controller deployed by the Helm chart is just in charge of communicating with Spacelift, and creating these runs.

When you deploy a pool using the operator, you define a `WorkerPool` custom resource. That resource defines information about your pool, for example the number of `Workers` you want to run concurrently, but also allows you to customize the configuration of the Pods that are created for each run (check out [our docs](https://docs.spacelift.io/concepts/worker-pools#configuration) for the full set of configuration options).

So I think probably what you want to do here is configure the Pods so that they create a side-car container running the Docker daemon, similar to what the old chart did. Frustratingly we don't currently allow additional side-car containers to be defined, although I don't think it would be a big deal for us to implement.

I think until then the only option would be to mount in the host's socket, like you mention. That should definitely be possible since you can define custom volume mounts right now, and adjust the security context of the Pods.

The TL;DR; of this is that making changes to the Helm chart wouldn't actually help here - what we actually need to change is to update the `WorkerPool` CRD to allow additional containers to be defined.

I'm going to close this issue for now since it's not quite related to the Helm chart, and open an internal development ticket to investigating adding this support.

**Comment by adamconnelly** [comment]
Just wanted to post an update on this - my colleague @eliecharra has updated the Kubernetes operator to support custom side-car containers. There's an open PR for our documentation to add information about this that will be merged shortly here: https://github.com/spacelift-io/user-documentation/pull/528.

And here's an example of how the new `additionalSidecarContainers` option can be used to expose a Docker daemon to the worker:

```yaml
apiVersion: workers.spacelift.io/v1beta1
kind: WorkerPool
metadata:
  name: test-workerpool
spec:
  poolSize: 2
  pod:
    workerContainer:
      env:
        - name: DOCKER_HOST
          value: tcp://localhost:2375
    additionalSidecarContainers:
      - image: docker:dind
        name: docker
        securityContext:
          privileged: true
        command:
          - docker-init
          - "--"
          - dockerd
          - "--host"
          - tcp://127.0.0.1:2375
```

The update is available in v0.0.9 of the controller (v0.20.0 of the Helm chart).

Thanks @eliecharra! :heart:
